hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java

```diff
@@ -164,8 +165,12 @@ private DataSerializable readInternal(ObjectDataInput in, Class<?> aClass)
                 in.readByte();
             }
 
-            ds.readData(in);
-            return ds;
+            if (ds instanceof IdentifiedDataSerializer) {
```

Comment:
current implementation does not protect against passing `aClass` for QDT. That would violate the enumness of the class. 

If we are fine with additional price of `instanceOf` (are we?) then the previous solution with `readResolve` (+ interface for it) was much less convoluted. The only downside is that we would create empty QDT instance and throw it out after `readResolve`.

Reply:
To fix that extra allocation we could instead add something like
```java
interface ResolvingIdentifiedDataSerializableFactory extends DataSerializableFactory {
    DataSerializable readOrResolve(int factoryId, int classId, @Nullable DataSerializable instance);
}
```
but that would require a few more changes.

Reply:
Important requirement is also that sql catalog should be available and readable during RU both on old and new members.

Reply:
Hotspot uses fast subtype checking algorithm[1]. `instance of` is pretty fast.
.
[1] -> https://dl.acm.org/doi/10.1145/583810.583821

Reply:
@burakgok could you elaborate more about backward compatibility? `SqlCatalogCompatibilityTest` fails with these changes. How it will work with:

1. Upgrade from 5.3 with existing SQL mappings (not necessarily referencing UDTs)
2. RU (eg. SqlCatalogCompatibilityTest) - during RU queries should work on a subset of the cluster
3. JDBC driver (OS and EE flavor)
4. MC (does it deserialize QDT in any case?)

Reply:
@Fly-Style Yes, `instanceof` is pretty fast, but as Krzysztof [pointed out][1], neither `HasSerializer` nor `IdentifiedDataSerializer` were really `IdentifiedDataSerializable`s. Imitating an interface by providing empty implementations is a strong indicator of poor design. That's why I deleted it. The current solution (5f47d406, hazelcast/hazelcast-enterprise@83d076c3) has the same performance impact.

@k-jamroz
1. `EnterpriseDataSerializableSerializer` writes the cluster version to the message, and when reading the message, it ensures that `in.getVersion() == versionReadFromMessage`. Thus, persisted `QueryDataType`s can be read by enterprise instances.
2. With [the introduction of `VersionedIdentifiedDataSerializable`][2], `SqlCatalogCompatibilityTest` started working again. It proves that queries will work during rolling upgrade.
3. JDBC drivers will receive `SqlTypeName` instead of `QueryDataType`. See `SqlColumnMetadata#type`.
4. Management Center seems to receive JSON messages, and maps them to objects with `DTO` suffix using Jackson. I encountered some non-primitive fields such as `JobInfoDTO#config`, which is of type `JobConfig`, but `QueryDataType` is not declared anywhere. If a serialized `QueryDataType` is received by Management Center somehow, MC won't be able to deserialize it just like an open-source instance.

[1]: https://github.com/hazelcast/hazelcast/pull/26056#discussion_r1412264037
[2]: https://github.com/hazelcast/hazelcast/commit/5f47d406acda1575f51528c0f486960d30e91461